File: go/kem.go
package topayz512

import (
	"crypto/rand"
	"crypto/sha256"
	"errors"
	"sync"
	"time"
)

// KEM (Key Encapsulation Mechanism) operations for TOPAY-Z512 with optimizations

// KEMResult represents the result of key encapsulation
type KEMResult struct {
	Ciphertext    []byte    `json:"ciphertext"`
	SharedSecret  []byte    `json:"shared_secret"`
	Timestamp     time.Time `json:"timestamp"`
	KeySize       uint32    `json:"key_size"`
	SecurityLevel uint32    `json:"security_level"`
}

// KEMDecryptResult represents the result of key decapsulation
type KEMDecryptResult struct {
	SharedSecret []byte    `json:"shared_secret"`
	IsValid      bool      `json:"is_valid"`
	Timestamp    time.Time `json:"timestamp"`
	KeySize      uint32    `json:"key_size"`
}

// Encapsulate performs key encapsulation with the public key using optimizations
func Encapsulate(publicKey *PublicKey) (*KEMResult, error) {
	if publicKey == nil {
		return nil, errors.New("public key cannot be nil")
	}

	// Validate public key
	if !IsValidPublicKey(*publicKey) {
		return nil, errors.New("invalid public key")
	}

	// Generate shared secret using pooled buffer
	sharedSecret := GetBuffer(SharedSecretSize)
	if _, err := rand.Read(sharedSecret); err != nil {
		PutBuffer(sharedSecret)
		return nil, err
	}

	// Create ciphertext using optimized encryption
	ciphertext, err := encryptSharedSecret(sharedSecret, publicKey)
	if err != nil {
		PutBuffer(sharedSecret)
		return nil, err
	}

	return &KEMResult{
		Ciphertext:    ciphertext,
		SharedSecret:  sharedSecret,
		Timestamp:     time.Now(),
		KeySize:       uint32(len(sharedSecret)),
		SecurityLevel: SecurityLevel,
	}, nil
}

// Decapsulate performs key decapsulation with the private key using optimizations
func Decapsulate(ciphertext []byte, privateKey *PrivateKey) (*KEMDecryptResult, error) {
	if privateKey == nil {
		return nil, errors.New("private key cannot be nil")
	}

	if len(ciphertext) == 0 {
		return nil, errors.New("ciphertext cannot be empty")
	}

	// Validate private key
	if !IsValidPrivateKey(*privateKey) {
		return nil, errors.New("invalid private key")
	}

	// Decrypt shared secret using optimized decryption
	sharedSecret, err := decryptSharedSecret(ciphertext, privateKey)
	if err != nil {
		return &KEMDecryptResult{
			SharedSecret: nil,
			IsValid:      false,
			Timestamp:    time.Now(),
			KeySize:      0,
		}, err
	}

	return &KEMDecryptResult{
		SharedSecret: sharedSecret,
		IsValid:      true,
		Timestamp:    time.Now(),
		KeySize:      uint32(len(sharedSecret)),
	}, nil
}

// encryptSharedSecret encrypts the shared secret with the public key using optimizations
func encryptSharedSecret(sharedSecret []byte, publicKey *PublicKey) ([]byte, error) {
	// Use pooled buffer for ciphertext
	ciphertext := GetBuffer(CiphertextSize)

	// Combine shared secret with public key data using SIMD operations
	keyData := publicKey.Bytes()

	// Use optimized XOR operation with SIMD when available
	if simdCaps.SSE2 && len(sharedSecret) >= 16 && len(keyData) >= len(sharedSecret) {
		VectorizedXOR(ciphertext[:len(sharedSecret)], sharedSecret, keyData[:len(sharedSecret)])
	} else {
		// Fallback to scalar XOR
		for i := 0; i < len(sharedSecret) && i < len(keyData); i++ {
			ciphertext[i] = sharedSecret[i] ^ keyData[i]
		}
	}

	// Add additional encryption layers
	hash := ComputeHash(ciphertext[:len(sharedSecret)])
	copy(ciphertext[len(sharedSecret):], hash[:])

	// Pad to full ciphertext size with secure random data
	if len(ciphertext) > len(sharedSecret)+HashSize {
		padding := ciphertext[len(sharedSecret)+HashSize:]
		if _, err := rand.Read(padding); err != nil {
			PutBuffer(ciphertext)
			return nil, err
		}
	}

	return ciphertext, nil
}

// decryptSharedSecret decrypts the shared secret with the private key using optimizations
func decryptSharedSecret(ciphertext []byte, privateKey *PrivateKey) ([]byte, error) {
	if len(ciphertext) < SharedSecretSize+HashSize {
		return nil, errors.New("ciphertext too short")
	}

	// Extract encrypted shared secret
	encryptedSecret := ciphertext[:SharedSecretSize]
	expectedHash := ciphertext[SharedSecretSize : SharedSecretSize+HashSize]

	// Use pooled buffer for decrypted secret
	sharedSecret := GetBuffer(SharedSecretSize)

	// Decrypt using private key with SIMD optimization
	keyData := privateKey.Bytes()

	if simdCaps.SSE2 && len(encryptedSecret) >= 16 && len(keyData) >= len(encryptedSecret) {
		VectorizedXOR(sharedSecret, encryptedSecret, keyData[:len(encryptedSecret)])
	} else {
		// Fallback to scalar XOR
		for i := 0; i < len(encryptedSecret) && i < len(keyData); i++ {
			sharedSecret[i] = encryptedSecret[i] ^ keyData[i]
		}
	}

	// Verify integrity using constant-time comparison
	computedHash := ComputeHash(encryptedSecret)
	if !ConstantTimeEqual(expectedHash, computedHash[:HashSize]) {
		PutBuffer(sharedSecret)
		return nil, errors.New("integrity check failed")
	}

	return sharedSecret, nil
}

// KEMKeyGen generates a new KEM key pair
func KEMKeyGen() (KEMPublicKey, KEMSecretKey, error) {
	// Generate random secret key
	secretBytes, err := SecureRandom(KEMSecretKeySize)
	if err != nil {
		return KEMPublicKey{}, KEMSecretKey{}, err
	}

	var secretKey KEMSecretKey
	copy(secretKey[:], secretBytes)

	// Derive public key from secret key
	publicKey := deriveKEMPublicKey(secretKey)

	return publicKey, secretKey, nil
}

// deriveKEMPublicKey derives a KEM public key from a secret key
func deriveKEMPublicKey(secretKey KEMSecretKey) KEMPublicKey {
	// Use SHA-256 based key derivation for KEM public key
	hasher := sha256.New()
	hasher.Write(secretKey[:])
	hasher.Write([]byte("TOPAY-Z512-KEM-PUBLIC-KEY"))

	firstHash := hasher.Sum(nil)

	// Second round for additional security
	hasher.Reset()
	hasher.Write(firstHash)
	hasher.Write(secretKey[:])
	hasher.Write([]byte("KEM-DERIVATION-ROUND-2"))
	secondHash := hasher.Sum(nil)

	var publicKey KEMPublicKey
	copy(publicKey[:], secondHash)

	return publicKey
}

// KEMEncapsulate encapsulates a shared secret using the public key
func KEMEncapsulate(publicKey KEMPublicKey) (Ciphertext, SharedSecret, error) {
	// Generate random ephemeral key
	ephemeralBytes, err := SecureRandom(32)
	if err != nil {
		return Ciphertext{}, SharedSecret{}, err
	}

	// Derive shared secret from ephemeral key and public key
	hasher := sha256.New()
	hasher.Write(ephemeralBytes)
	hasher.Write(publicKey[:])
	hasher.Write([]byte("TOPAY-Z512-KEM-SHARED-SECRET"))

	sharedSecretHash := hasher.Sum(nil)

	var sharedSecret SharedSecret
	copy(sharedSecret[:], sharedSecretHash)

	// Create ciphertext by encrypting ephemeral key with public key
	ciphertext := createCiphertext(ephemeralBytes, publicKey)

	return ciphertext, sharedSecret, nil
}

// KEMDecapsulate decapsulates the shared secret using the secret key
func KEMDecapsulate(secretKey KEMSecretKey, ciphertext Ciphertext) (SharedSecret, error) {
	// Derive public key from secret key for verification
	publicKey := deriveKEMPublicKey(secretKey)

	// Decrypt ephemeral key from ciphertext
	ephemeralBytes, err := decryptCiphertext(ciphertext, secretKey)
	if err != nil {
		return SharedSecret{}, ErrDecapsulationFailed
	}

	// Derive shared secret from ephemeral key and public key
	hasher := sha256.New()
	hasher.Write(ephemeralBytes)
	hasher.Write(publicKey[:])
	hasher.Write([]byte("TOPAY-Z512-KEM-SHARED-SECRET"))

	sharedSecretHash := hasher.Sum(nil)

	var sharedSecret SharedSecret
	copy(sharedSecret[:], sharedSecretHash)

	return sharedSecret, nil
}

// createCiphertext creates a ciphertext by encrypting ephemeral key with public key
func createCiphertext(ephemeralKey []byte, publicKey KEMPublicKey) Ciphertext {
	// Simple XOR-based encryption for demonstration
	// In production, use proper lattice-based encryption
	hasher := sha256.New()
	hasher.Write(publicKey[:])
	hasher.Write([]byte("TOPAY-Z512-KEM-ENCRYPTION-KEY"))

	encryptionKey := hasher.Sum(nil)

	var ciphertext Ciphertext

	// XOR ephemeral key with derived encryption key
	for i := 0; i < len(ephemeralKey) && i < CiphertextSize; i++ {
		ciphertext[i] = ephemeralKey[i] ^ encryptionKey[i%len(encryptionKey)]
	}

	// Fill remaining bytes with hash of the encrypted portion
	if len(ephemeralKey) < CiphertextSize {
		hasher.Reset()
		hasher.Write(ciphertext[:len(ephemeralKey)])
		hasher.Write(publicKey[:])
		fillHash := hasher.Sum(nil)

		for i := len(ephemeralKey); i < CiphertextSize; i++ {
			ciphertext[i] = fillHash[i%len(fillHash)]
		}
	}

	return ciphertext
}

// decryptCiphertext decrypts the ciphertext to recover ephemeral key
func decryptCiphertext(ciphertext Ciphertext, secretKey KEMSecretKey) ([]byte, error) {
	// Derive public key from secret key
	publicKey := deriveKEMPublicKey(secretKey)

	// Derive encryption key
	hasher := sha256.New()
	hasher.Write(publicKey[:])
	hasher.Write([]byte("TOPAY-Z512-KEM-ENCRYPTION-KEY"))

	encryptionKey := hasher.Sum(nil)

	// Decrypt ephemeral key (assuming 32 bytes)
	ephemeralKey := make([]byte, 32)
	for i := 0; i < len(ephemeralKey); i++ {
		ephemeralKey[i] = ciphertext[i] ^ encryptionKey[i%len(encryptionKey)]
	}

	return ephemeralKey, nil
}

// Batch KEM operations

// BatchKEMResult represents the result of batch KEM operations
type BatchKEMResult struct {
	Index        int
	PublicKey    KEMPublicKey
	SecretKey    KEMSecretKey
	Ciphertext   Ciphertext
	SharedSecret SharedSecret
	Error        error
}

// BatchKEMKeyGen generates multiple KEM key pairs in parallel
func BatchKEMKeyGen(count int) ([]KEMPublicKey, []KEMSecretKey, error) {
	if count <= 0 {
		return nil, nil, ErrInvalidFragmentCount
	}

	publicKeys := make([]KEMPublicKey, count)
	secretKeys := make([]KEMSecretKey, count)

	// Use optimal number of goroutines
	numWorkers := OptimalThreadCount()
	if numWorkers > count {
		numWorkers = count
	}

	// Channel for work distribution
	workChan := make(chan int, count)
	resultChan := make(chan BatchKEMResult, count)

	// Start workers
	var wg sync.WaitGroup
	for i := 0; i < numWorkers; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for index := range workChan {
				publicKey, secretKey, err := KEMKeyGen()
				resultChan <- BatchKEMResult{
					Index:     index,
					PublicKey: publicKey,
					SecretKey: secretKey,
					Error:     err,
				}
			}
		}()
	}

	// Send work
	go func() {
		for i := 0; i < count; i++ {
			workChan <- i
		}
		close(workChan)
	}()

	// Wait for workers to complete
	go func() {
		wg.Wait()
		close(resultChan)
	}()

	// Collect results
	for result := range resultChan {
		if result.Error != nil {
			return nil, nil, result.Error
		}
		publicKeys[result.Index] = result.PublicKey
		secretKeys[result.Index] = result.SecretKey
	}

	return publicKeys, secretKeys, nil
}

// BatchKEMEncapsulate performs multiple encapsulations in parallel
func BatchKEMEncapsulate(publicKeys []KEMPublicKey) ([]Ciphertext, []SharedSecret, error) {
	if len(publicKeys) == 0 {
		return nil, nil, ErrEmptyData
	}

	ciphertexts := make([]Ciphertext, len(publicKeys))
	sharedSecrets := make([]SharedSecret, len(publicKeys))

	// Use optimal number of goroutines
	numWorkers := OptimalThreadCount()
	if numWorkers > len(publicKeys) {
		numWorkers = len(publicKeys)
	}

	// Channel for work distribution
	workChan := make(chan int, len(publicKeys))
	resultChan := make(chan BatchKEMResult, len(publicKeys))

	// Start workers
	var wg sync.WaitGroup
	for i := 0; i < numWorkers; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for index := range workChan {
				ciphertext, sharedSecret, err := KEMEncapsulate(publicKeys[index])
				resultChan <- BatchKEMResult{
					Index:        index,
					Ciphertext:   ciphertext,
					SharedSecret: sharedSecret,
					Error:        err,
				}
			}
		}()
	}

	// Send work
	go func() {
		for i := range publicKeys {
			workChan <- i
		}
		close(workChan)
	}()

	// Wait for workers to complete
	go func() {
		wg.Wait()
		close(resultChan)
	}()

	// Collect results
	for result := range resultChan {
		if result.Error != nil {
			return nil, nil, result.Error
		}
		ciphertexts[result.Index] = result.Ciphertext
		sharedSecrets[result.Index] = result.SharedSecret
	}

	return ciphertexts, sharedSecrets, nil
}

// BatchKEMDecapsulate performs multiple decapsulations in parallel
func BatchKEMDecapsulate(secretKeys []KEMSecretKey, ciphertexts []Ciphertext) ([]SharedSecret, error) {
	if len(secretKeys) != len(ciphertexts) {
		return nil, ErrInvalidFragmentCount
	}

	if len(secretKeys) == 0 {
		return nil, ErrEmptyData
	}

	sharedSecrets := make([]SharedSecret, len(secretKeys))

	// Use optimal number of goroutines
	numWorkers := OptimalThreadCount()
	if numWorkers > len(secretKeys) {
		numWorkers = len(secretKeys)
	}

	// Channel for work distribution
	workChan := make(chan int, len(secretKeys))
	resultChan := make(chan BatchKEMResult, len(secretKeys))

	// Start workers
	var wg sync.WaitGroup
	for i := 0; i < numWorkers; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for index := range workChan {
				sharedSecret, err := KEMDecapsulate(secretKeys[index], ciphertexts[index])
				resultChan <- BatchKEMResult{
					Index:        index,
					SharedSecret: sharedSecret,
					Error:        err,
				}
			}
		}()
	}

	// Send work
	go func() {
		for i := range secretKeys {
			workChan <- i
		}
		close(workChan)
	}()

	// Wait for workers to complete
	go func() {
		wg.Wait()
		close(resultChan)
	}()

	// Collect results
	for result := range resultChan {
		if result.Error != nil {
			return nil, result.Error
		}
		sharedSecrets[result.Index] = result.SharedSecret
	}

	return sharedSecrets, nil
}

// KEM validation and utilities

// VerifyKEMKeyPair verifies that a KEM public and secret key form a valid pair
func VerifyKEMKeyPair(publicKey KEMPublicKey, secretKey KEMSecretKey) bool {
	derivedPublic := deriveKEMPublicKey(secretKey)
	return ConstantTimeEqual(publicKey[:], derivedPublic[:])
}

// IsValidKEMPublicKey checks if a KEM public key is valid
func IsValidKEMPublicKey(publicKey KEMPublicKey) bool {
	var zero KEMPublicKey
	return !ConstantTimeEqual(publicKey[:], zero[:])
}

// IsValidKEMSecretKey checks if a KEM secret key is valid
func IsValidKEMSecretKey(secretKey KEMSecretKey) bool {
	var zero KEMSecretKey
	return !ConstantTimeEqual(secretKey[:], zero[:])
}

// IsValidCiphertext checks if a ciphertext is valid
func IsValidCiphertext(ciphertext Ciphertext) bool {
	var zero Ciphertext
	return !ConstantTimeEqual(ciphertext[:], zero[:])
}

// IsValidSharedSecret checks if a shared secret is valid
func IsValidSharedSecret(sharedSecret SharedSecret) bool {
	var zero SharedSecret
	return !ConstantTimeEqual(sharedSecret[:], zero[:])
}

// KEMEqual compares KEM components for equality

// KEMPublicKeyEqual compares two KEM public keys for equality
func KEMPublicKeyEqual(pk1, pk2 KEMPublicKey) bool {
	return ConstantTimeEqual(pk1[:], pk2[:])
}

// KEMSecretKeyEqual compares two KEM secret keys for equality
func KEMSecretKeyEqual(sk1, sk2 KEMSecretKey) bool {
	return ConstantTimeEqual(sk1[:], sk2[:])
}

// CiphertextEqual compares two ciphertexts for equality
func CiphertextEqual(ct1, ct2 Ciphertext) bool {
	return ConstantTimeEqual(ct1[:], ct2[:])
}

// SharedSecretEqual compares two shared secrets for equality
func SharedSecretEqual(ss1, ss2 SharedSecret) bool {
	return ConstantTimeEqual(ss1[:], ss2[:])
}

// Secure erasure functions

// SecureEraseKEMSecretKey securely erases a KEM secret key from memory
func SecureEraseKEMSecretKey(secretKey *KEMSecretKey) {
	SecureZero(secretKey[:])
}

// SecureEraseSharedSecret securely erases a shared secret from memory
func SecureEraseSharedSecret(sharedSecret *SharedSecret) {
	SecureZero(sharedSecret[:])
}

// SecureEraseKEMKeyPair securely erases a KEM key pair from memory
func SecureEraseKEMKeyPair(keyPair *KEMKeyPair) {
	SecureZero(keyPair.Public[:])
	SecureZero(keyPair.Secret[:])
}

// KEM performance benchmarking

// KEMBenchmark represents KEM performance metrics
type KEMBenchmark struct {
	KeyGenPerSec      float64
	EncapsulatePerSec float64
	DecapsulatePerSec float64
	BatchSpeedupRatio float64
	AvgLatencyMs      float64
}

// BenchmarkKEM measures KEM performance
func BenchmarkKEM(iterations int) KEMBenchmark {
	// Key generation benchmark
	start := time.Now()
	for i := 0; i < iterations; i++ {
		_, _, _ = KEMKeyGen()
	}
	keyGenDuration := time.Since(start)

	// Generate test key pair
	publicKey, secretKey, _ := KEMKeyGen()

	// Encapsulation benchmark
	start = time.Now()
	var ciphertext Ciphertext
	for i := 0; i < iterations; i++ {
		ciphertext, _, _ = KEMEncapsulate(publicKey)
	}
	encapDuration := time.Since(start)

	// Decapsulation benchmark
	start = time.Now()
	for i := 0; i < iterations; i++ {
		_, _ = KEMDecapsulate(secretKey, ciphertext)
	}
	decapDuration := time.Since(start)

	// Batch benchmark
	start = time.Now()
	_, _, _ = BatchKEMKeyGen(iterations)
	batchDuration := time.Since(start)

	keyGenPerSec := float64(iterations) / keyGenDuration.Seconds()
	encapsulatePerSec := float64(iterations) / encapDuration.Seconds()
	decapsulatePerSec := float64(iterations) / decapDuration.Seconds()
	batchSpeedupRatio := keyGenDuration.Seconds() / batchDuration.Seconds()
	avgLatencyMs := keyGenDuration.Seconds() * 1000 / float64(iterations)

	return KEMBenchmark{
		KeyGenPerSec:      keyGenPerSec,
		EncapsulatePerSec: encapsulatePerSec,
		DecapsulatePerSec: decapsulatePerSec,
		BatchSpeedupRatio: batchSpeedupRatio,
		AvgLatencyMs:      avgLatencyMs,
	}
}

// Advanced KEM operations

// KEMWithContext performs KEM operations with additional context data
func KEMWithContext(publicKey KEMPublicKey, context []byte) (Ciphertext, SharedSecret, error) {
	// Generate random ephemeral key
	ephemeralBytes, err := SecureRandom(32)
	if err != nil {
		return Ciphertext{}, SharedSecret{}, err
	}

	// Derive shared secret with context
	hasher := sha256.New()
	hasher.Write(ephemeralBytes)
	hasher.Write(publicKey[:])
	hasher.Write(context)
	hasher.Write([]byte("TOPAY-Z512-KEM-CONTEXT-SECRET"))

	sharedSecretHash := hasher.Sum(nil)

	var sharedSecret SharedSecret
	copy(sharedSecret[:], sharedSecretHash)

	// Create ciphertext
	ciphertext := createCiphertext(ephemeralBytes, publicKey)

	return ciphertext, sharedSecret, nil
}

// KEMDecapsulateWithContext decapsulates with additional context data
func KEMDecapsulateWithContext(secretKey KEMSecretKey, ciphertext Ciphertext, context []byte) (SharedSecret, error) {
	// Derive public key from secret key
	publicKey := deriveKEMPublicKey(secretKey)

	// Decrypt ephemeral key from ciphertext
	ephemeralBytes, err := decryptCiphertext(ciphertext, secretKey)
	if err != nil {
		return SharedSecret{}, ErrDecapsulationFailed
	}

	// Derive shared secret with context
	hasher := sha256.New()
	hasher.Write(ephemeralBytes)
	hasher.Write(publicKey[:])
	hasher.Write(context)
	hasher.Write([]byte("TOPAY-Z512-KEM-CONTEXT-SECRET"))

	sharedSecretHash := hasher.Sum(nil)

	var sharedSecret SharedSecret
	copy(sharedSecret[:], sharedSecretHash)

	return sharedSecret, nil
}

File: js/src/kem.ts
/**
 * Key Encapsulation Mechanism (KEM) for TOPAY-Z512
 */

import {
  KEM_PUBLIC_KEY_SIZE,
  KEM_SECRET_KEY_SIZE,
  CIPHERTEXT_SIZE,
  // SHARED_SECRET_SIZE,
  KEM_SHARED_SECRET_SIZE,
  KEMKeyPair,
  KEMPublicKey,
  KEMSecretKey,
  Ciphertext,
  SharedSecret,
  EncapsulationResult,
  DecapsulationFailedError
} from './index';
import {
  secureRandom,
  validateSize,
  constantTimeEqual,
  secureZero,
  copyBytes,
  xorBytes
} from './utils';
import { computeHash /* , computeHmac */ } from './hash';

// Key pair cache for improved performance
const keyPairCache = new Map<
  string,
  { publicKey: Uint8Array; secretKey: Uint8Array; timestamp: number }
>();
const KEY_CACHE_TTL = 300000; // 5 minutes
const KEY_CACHE_MAX_SIZE = 100;

/**
 * Clears expired entries from the key pair cache
 */
function cleanupKeyCache(): void {
  const now = Date.now();
  for (const [key, value] of Array.from(keyPairCache.entries())) {
    if (now - value.timestamp > KEY_CACHE_TTL) {
      keyPairCache.delete(key);
    }
  }
}

/**
 * Optimized KEM key generation with optional caching
 * @param useCache - Whether to use caching (default: false for security)
 * @returns Promise resolving to a new KEM key pair
 */
export async function kemKeyGen(useCache: boolean = false): Promise<KEMKeyPair> {
  if (useCache) {
    // Clean up expired entries periodically
    if (keyPairCache.size > 0 && Math.random() < 0.1) {
      cleanupKeyCache();
    }

    // For demonstration purposes, we'll cache based on a simple key
    // In production, this should be more sophisticated
    const cacheKey = 'default';
    const cached = keyPairCache.get(cacheKey);

    if (cached && Date.now() - cached.timestamp < KEY_CACHE_TTL) {
      return {
        publicKey: new Uint8Array(cached.publicKey),
        secretKey: new Uint8Array(cached.secretKey)
      };
    }
  }

  // Generate secret key from secure random bytes
  const secretKey = await secureRandom(KEM_SECRET_KEY_SIZE);

  // Derive public key from secret key
  const publicKey = await deriveKEMPublicKey(secretKey);

  if (useCache) {
    // Manage cache size
    if (keyPairCache.size >= KEY_CACHE_MAX_SIZE) {
      const firstKey = keyPairCache.keys().next().value;
      if (firstKey !== undefined) {
        keyPairCache.delete(firstKey);
      }
    }

    keyPairCache.set('default', {
      publicKey: new Uint8Array(publicKey),
      secretKey: new Uint8Array(secretKey),
      timestamp: Date.now()
    });
  }

  return {
    publicKey,
    secretKey
  };
}

/**
 * Derives a KEM public key from a secret key
 * @param secretKey - Secret key to derive from
 * @returns Promise resolving to the derived public key
 * @throws InvalidKeySizeError if secret key size is invalid
 */
export async function deriveKEMPublicKey(secretKey: KEMSecretKey): Promise<KEMPublicKey> {
  validateSize(secretKey, KEM_SECRET_KEY_SIZE, 'KEM secret key');

  // Use hash-based public key derivation
  const publicKey = await computeHash(secretKey);
  validateSize(publicKey, KEM_PUBLIC_KEY_SIZE, 'KEM public key');

  return publicKey;
}

/**
 * Encapsulates a shared secret using a public key
 * @param publicKey - KEM public key for encapsulation
 * @returns Promise resolving to encapsulation result (ciphertext and shared secret)
 * @throws InvalidKeySizeError if public key size is invalid
 */
export async function kemEncapsulate(publicKey: KEMPublicKey): Promise<EncapsulationResult> {
  validateSize(publicKey, KEM_PUBLIC_KEY_SIZE, 'KEM public key');

  // Generate ephemeral key material
  const ephemeralKey = await secureRandom(32);

  // Derive shared secret using key derivation
  const keyMaterial = new Uint8Array(publicKey.length + ephemeralKey.length);
  keyMaterial.set(publicKey);
  keyMaterial.set(ephemeralKey, publicKey.length);

  const fullHash = await computeHash(keyMaterial);
  // Truncate to the expected KEM shared secret size
  const sharedSecret = fullHash.slice(0, KEM_SHARED_SECRET_SIZE);
  validateSize(sharedSecret, KEM_SHARED_SECRET_SIZE, 'shared secret');

  // Create ciphertext by encrypting ephemeral key with public key
  const ciphertext = await encryptEphemeralKey(ephemeralKey, publicKey);
  validateSize(ciphertext, CIPHERTEXT_SIZE, 'ciphertext');

  return {
    ciphertext,
    sharedSecret
  };
}

/**
 * Decapsulates a shared secret using a secret key and ciphertext
 * @param secretKey - KEM secret key for decapsulation
 * @param ciphertext - Ciphertext to decapsulate
 * @returns Promise resolving to the shared secret
 * @throws InvalidKeySizeError if key or ciphertext size is invalid
 * @throws DecapsulationFailedError if decapsulation fails
 */
export async function kemDecapsulate(
  secretKey: KEMSecretKey,
  ciphertext: Ciphertext
): Promise<SharedSecret> {
  validateSize(secretKey, KEM_SECRET_KEY_SIZE, 'KEM secret key');
  validateSize(ciphertext, CIPHERTEXT_SIZE, 'ciphertext');

  try {
    // Derive public key from secret key
    const publicKey = await deriveKEMPublicKey(secretKey);

    // Decrypt ephemeral key from ciphertext
    const ephemeralKey = await decryptEphemeralKey(ciphertext, secretKey);

    // Derive shared secret using the same process as encapsulation
    const keyMaterial = new Uint8Array(publicKey.length + ephemeralKey.length);
    keyMaterial.set(publicKey);
    keyMaterial.set(ephemeralKey, publicKey.length);

    const fullHash = await computeHash(keyMaterial);
    // Truncate to the expected KEM shared secret size
    const sharedSecret = fullHash.slice(0, KEM_SHARED_SECRET_SIZE);
    validateSize(sharedSecret, KEM_SHARED_SECRET_SIZE, 'shared secret');

    return sharedSecret;
  } catch (error) {
    throw new DecapsulationFailedError();
  }
}

/**
 * Performs batch KEM key generation
 * @param count - Number of key pairs to generate
 * @returns Promise resolving to array of KEM key pairs
 */
export async function batchKEMKeyGen(count: number): Promise<KEMKeyPair[]> {
  if (count <= 0) {
    throw new Error('Count must be positive');
  }

  const promises: Promise<KEMKeyPair>[] = [];
  for (let i = 0; i < count; i++) {
    promises.push(kemKeyGen());
  }

  return Promise.all(promises);
}

/**
 * Performs batch KEM encapsulation
 * @param publicKeys - Array of public keys for encapsulation
 * @returns Promise resolving to array of encapsulation results
 */
export async function batchKEMEncapsulate(
  publicKeys: KEMPublicKey[]
): Promise<EncapsulationResult[]> {
  const promises = publicKeys.map(publicKey => kemEncapsulate(publicKey));
  return Promise.all(promises);
}

/**
 * Performs batch KEM decapsulation
 * @param operations - Array of {secretKey, ciphertext} pairs
 * @returns Promise resolving to array of shared secrets
 */
export async function batchKEMDecapsulate(
  operations: Array<{ secretKey: KEMSecretKey; ciphertext: Ciphertext }>
): Promise<SharedSecret[]> {
  const promises = operations.map(({ secretKey, ciphertext }) =>
    kemDecapsulate(secretKey, ciphertext)
  );
  return Promise.all(promises);
}

/**
 * Validates a KEM key pair
 * @param keyPair - KEM key pair to validate
 * @returns Promise resolving to true if valid
 */
export async function validateKEMKeyPair(keyPair: KEMKeyPair): Promise<boolean> {
  try {
    validateSize(keyPair.secretKey, KEM_SECRET_KEY_SIZE, 'KEM secret key');
    validateSize(keyPair.publicKey, KEM_PUBLIC_KEY_SIZE, 'KEM public key');

    const derivedPublic = await deriveKEMPublicKey(keyPair.secretKey);
    return constantTimeEqual(keyPair.publicKey, derivedPublic);
  } catch {
    return false;
  }
}

/**
 * Tests KEM operations with a key pair
 * @param keyPair - KEM key pair to test
 * @returns Promise resolving to true if operations succeed
 */
export async function testKEMOperations(keyPair: KEMKeyPair): Promise<boolean> {
  try {
    // Test encapsulation
    const { ciphertext, sharedSecret } = await kemEncapsulate(keyPair.publicKey);

    // Test decapsulation
    const decapsulatedSecret = await kemDecapsulate(keyPair.secretKey, ciphertext);

    // Verify shared secrets match
    return constantTimeEqual(sharedSecret, decapsulatedSecret);
  } catch {
    return false;
  }
}

/**
 * Securely erases a KEM key pair from memory
 * @param keyPair - KEM key pair to erase
 */
export function secureEraseKEMKeyPair(keyPair: KEMKeyPair): void {
  secureZero(keyPair.secretKey);
  secureZero(keyPair.publicKey);
}

/**
 * Creates a backup of a KEM key pair
 * @param keyPair - KEM key pair to backup
 * @returns New KEM key pair with copied data
 */
export function backupKEMKeyPair(keyPair: KEMKeyPair): KEMKeyPair {
  return {
    publicKey: copyBytes(keyPair.publicKey),
    secretKey: copyBytes(keyPair.secretKey)
  };
}

/**
 * Serializes a KEM key pair to JSON
 * @param keyPair - KEM key pair to serialize
 * @returns Serialized key pair as JSON string
 */
export function serializeKEMKeyPair(keyPair: KEMKeyPair): string {
  return JSON.stringify({
    publicKey: Array.from(keyPair.publicKey),
    secretKey: Array.from(keyPair.secretKey)
  });
}

/**
 * Deserializes a KEM key pair from JSON
 * @param serialized - Serialized key pair JSON string
 * @returns Deserialized KEM key pair
 * @throws Error if deserialization fails
 */
export function deserializeKEMKeyPair(serialized: string): KEMKeyPair {
  try {
    const data = JSON.parse(serialized);

    if (!data.publicKey || !data.secretKey) {
      throw new Error('Missing key data');
    }

    const publicKey = new Uint8Array(data.publicKey);
    const secretKey = new Uint8Array(data.secretKey);

    validateSize(publicKey, KEM_PUBLIC_KEY_SIZE, 'KEM public key');
    validateSize(secretKey, KEM_SECRET_KEY_SIZE, 'KEM secret key');

    return { publicKey, secretKey };
  } catch (error) {
    throw new Error(`Failed to deserialize KEM key pair: ${error}`);
  }
}

// Helper functions for encryption/decryption of ephemeral keys

/**
 * Encrypts ephemeral key material with public key
 * @param ephemeralKey - Ephemeral key to encrypt
 * @param publicKey - Public key for encryption
 * @returns Promise resolving to ciphertext
 */
async function encryptEphemeralKey(
  ephemeralKey: Uint8Array,
  publicKey: KEMPublicKey
): Promise<Ciphertext> {
  // Simple XOR-based encryption for demonstration
  // In production, use proper lattice-based encryption
  const keyStream = await computeHash(publicKey);
  const encrypted = xorBytes(ephemeralKey, keyStream.slice(0, ephemeralKey.length));

  // Pad to ciphertext size
  const ciphertext = new Uint8Array(CIPHERTEXT_SIZE);
  ciphertext.set(encrypted);

  return ciphertext;
}

/**
 * Decrypts ephemeral key material with secret key
 * @param ciphertext - Ciphertext to decrypt
 * @param secretKey - Secret key for decryption
 * @returns Promise resolving to ephemeral key
 */
async function decryptEphemeralKey(
  ciphertext: Ciphertext,
  secretKey: KEMSecretKey
): Promise<Uint8Array> {
  // Derive public key and decrypt
  const publicKey = await deriveKEMPublicKey(secretKey);
  const keyStream = await computeHash(publicKey);

  const encryptedPart = ciphertext.slice(0, 32); // Extract encrypted ephemeral key
  const ephemeralKey = xorBytes(encryptedPart, keyStream.slice(0, 32));

  return ephemeralKey;
}
